### ISSUE TITLE
Does the neighbor lookups offset the all-pairs?

### ISSUE THREAD

**Issue by ProNoobLi** [princeton-vl/RAFT#12]
Hi, I am curious about the operation of neighbor lookups applied on all-pairs correlation.
As I understand, each entry of all-pair cost volume indicates a correlation value between each vector in fmap1 and each vector in fmap2. 
However, neighbor lookups only capture a range of correlation values to be catenated into a feature map. Thus, all-pairs correlation is not fully utilized in the following steps, and the correlation is still computed according to the search range.
![image](https://user-images.githubusercontent.com/33815430/83613821-ec946300-a5b6-11ea-8dca-f1d631f74517.png)

For example, consider the corr at level 1 is: [48x128, 1, 48. 128 ]=[6144, 1, 48, 128] (ignoring the batch)
coords_lvl: [6144, 7, 7, 2] which represents that each pix in fmap1 has a (7, 7) coordinates of its neighbor. 
Then bilinear_sampler samples the all-pair cost volume according to coords_lvl.
Finally, the result corr: [48, 128, 49]. Each entry stores the correlation value between each feature in fmap1 and its 7x7 neighbor in fmap2.     
 Where is the all pair??????

**Comment by JiayuanSternLi** [comment]
Hi, 
To verify my opinion, I implement a very naive efficient approach, but the result becomes worse. 
![image](https://user-images.githubusercontent.com/33815430/83855309-7a548780-a74a-11ea-817d-fbbd857556a3.png)
Warped results from the predicted_flow result from the original code(all pairs). Such warped results are masked by the occlusion mask.
![image](https://user-images.githubusercontent.com/33815430/83855440-a6700880-a74a-11ea-8f69-54d0b5c29300.png)
Warped results from the predicted_flow result from my code(local neighbor search). Such warped results are masked by the occlusion mask.

I can see that my implementation is much worse than yours, while my implementation is based on my opinion in the post above. The code is attached here, and please feel free to point out what I misunderstood.
![image](https://user-images.githubusercontent.com/33815430/83855907-5b0a2a00-a74b-11ea-940e-a1aaafac7472.png)
![image](https://user-images.githubusercontent.com/33815430/83855944-665d5580-a74b-11ea-83e9-de7ba6f1d6cd.png)

**Comment by JiayuanSternLi** [comment]
Let me explain here.
For your all-pairs correlation. The method and implementation is straightforward and easy to understand. 
1. Let's assume the fmap1 and fmap2 [B, C, H, W] = [B, 128, 48, 128]
2. Get an all-pairs correlation by calculating a 6144*6144 matrix multiplication.
3. Downsample such matrix to get the corr_pyramid for 4 levels
4. Lookup the correlation values of a 7*7 neighbors in fmap2 surrounding each pix at fmap1 .
5. Cat four correlation feature maps [48, 128, 49]*4 and return a [48, 128, 196] feature.
From my understanding, though the all-pairs correlation is calculated, only 196 out of 6144 times dot product are utilized. Thus ideally, there exists an alternative implementation takes only O(NM) as  your paper mentioned.

**Comment by zachteed** [comment]
Hi, I'm having trouble following. But, here is some code that will do the job
`
class CorrBlock:
    def __init__(self, fmap1, fmap2, num_levels=4, radius=4):
        self.num_levels = num_levels
        self.radius = radius
        self.corr_pyramid = []

        self.fmap1 = fmap1
        self.fmap2 = fmap2

    def corr(self, fmap1, fmap2, coords):

        B, D, H, W = fmap2.shape
        fmap1 = fmap1.unsqueeze(dim=-1)
        fmap2 = fmap2.unsqueeze(dim=-1)

        # map grid coordinates to [-1,1]
        xgrid, ygrid = coords.split([1,1], dim=-1)
        xgrid = 2*xgrid/(W-1) - 1
        ygrid = 2*ygrid/(H-1) - 1
        zgrid = torch.zeros_like(xgrid) - 1
        grid = torch.cat([zgrid, xgrid, ygrid], dim=-1)

        fmapw = F.grid_sample(fmap2, grid, align_corners=True)

        corr = torch.sum(fmap1*fmapw, dim=1)
        return corr / torch.sqrt(torch.tensor(D).float())

    def __call__(self, coords):

        r = self.radius
        coords = coords.permute(0, 2, 3, 1)
        batch, h1, w1, _ = coords.shape

        fmap1 = self.fmap1
        fmap2 = self.fmap2

        out_pyramid = []
        for i in range(self.num_levels):
            dx = torch.linspace(-r, r, 2*r+1)
            dy = torch.linspace(-r, r, 2*r+1)
            delta = torch.stack(torch.meshgrid(dy, dx), axis=-1).to(coords.device)

            centroid_lvl = coords.reshape(batch, h1, w1, 1, 2) / 2**i
            coords_lvl = centroid_lvl + delta.view(-1, 2)

            corr = self.corr(fmap1, fmap2, coords_lvl)
            fmap2 = F.avg_pool2d(fmap2, 2, stride=2)
            out_pyramid.append(corr)

        out = torch.cat(out_pyramid, dim=-1)
        return out.permute(0, 3, 1, 2).contiguous().float()
`

**Comment by zachteed** [comment]
I think the problem with your code was that you weren't scaling the correlation features by 1.0/sqrt(D)

**Comment by JiayuanSternLi** [comment]
> Hi, I'm having trouble following. But, here is some code that will do the job. Just plug this in instead of the default CorrBlock. You can check that it gives exactly (to numerical precision) results as the default implementation. There is no need to loop over the neighboorhood, you can just use the built in grid sampler. Again, this is very slow, you will need to write a cuda implementation for it to be practical.
> 
> ```
> class CorrBlock:
>     def __init__(self, fmap1, fmap2, num_levels=4, radius=4):
>         self.num_levels = num_levels
>         self.radius = radius
>         self.corr_pyramid = []
> 
>         self.fmap1 = fmap1
>         self.fmap2 = fmap2
> 
>     def corr(self, fmap1, fmap2, coords):
> 
>         B, D, H, W = fmap2.shape
>         fmap1 = fmap1.unsqueeze(dim=-1)
>         fmap2 = fmap2.unsqueeze(dim=-1)
> 
>         # map grid coordinates to [-1,1]
>         xgrid, ygrid = coords.split([1,1], dim=-1)
>         xgrid = 2*xgrid/(W-1) - 1
>         ygrid = 2*ygrid/(H-1) - 1
>         zgrid = torch.zeros_like(xgrid) - 1
>         grid = torch.cat([zgrid, xgrid, ygrid], dim=-1)
> 
>         fmapw = F.grid_sample(fmap2, grid, align_corners=True)
> 
>         corr = torch.sum(fmap1*fmapw, dim=1)
>         return corr / torch.sqrt(torch.tensor(D).float())
> 
>     def __call__(self, coords):
> 
>         r = self.radius
>         coords = coords.permute(0, 2, 3, 1)
>         batch, h1, w1, _ = coords.shape
> 
>         fmap1 = self.fmap1
>         fmap2 = self.fmap2
> 
>         out_pyramid = []
>         for i in range(self.num_levels):
>             dx = torch.linspace(-r, r, 2*r+1)
>             dy = torch.linspace(-r, r, 2*r+1)
>             delta = torch.stack(torch.meshgrid(dy, dx), axis=-1).to(coords.device)
> 
>             centroid_lvl = coords.reshape(batch, h1, w1, 1, 2) / 2**i
>             coords_lvl = centroid_lvl + delta.view(-1, 2)
> 
>             corr = self.corr(fmap1, fmap2, coords_lvl)
>             fmap2 = F.avg_pool2d(fmap2, 2, stride=2)
>             out_pyramid.append(corr)
> 
>         out = torch.cat(out_pyramid, dim=-1)
>         return out.permute(0, 3, 1, 2).contiguous().float()
> ```

Thank you for the code. It's straightforward to understand.
Here is my question: anyway you grid sample the neighbor of each feature, thus all-pairs correlation isn't utilized in the following step, right? I think it still be the local correlation search and the search range is determined by lookup radius.

**Comment by zachteed** [comment]
At the lowest pyramid level, the vector returned by the lookup operator at a single point is a function of 72x72 pairs (576x576 in pixels). Yes, the search range is determined by lookup radius, but the search range covers the majority, if not all, of the image.

The only reason the efficient implementation exists is because we use the dot product for correlation, and average pooling for aggregation. Both these operations are linear. Swap either of these operators with a non-linear function (i.e. L1 distance, max pooling), and our "efficient implementation" no longer exists. 

The code I provided above is ~8x slower than our all-pairs implementation. So even when an implementation with lower complexity exists, there are practical advantages to our formulation.

**Comment by JiayuanSternLi** [comment]
Great! I see. It sounds like the receptive area at low level covers a wider range over the high level, but here all information is not from CNN but correlation.

PWC-net and flownet downsample the feature map, but they use local search at each level, while yours uses all-pairs at each level. 

It makes sense! A new idea. Congrats
